gslib/gcs_json_api.py

```diff
@@ -174,6 +175,34 @@
 MTLS_HOST = 'storage.mtls.googleapis.com'
 
 
+def _GetPatchedFinalizeTransferUrl(original_function, correct_netloc):
+  """Patch the Apitools client to always respect custom endpoints.
+
+  Complex uploads and downloads have cases where they initially take our
+  custom endpoints but ignore them later. This function
+  rewrites the URL host to the desired custom one before requests.
+
+  Args:
+    original_function: The original client FinalizeTransferUrl method.
+      We still want to run this, so we don't fall out of sync with the Apitools
+      dependency.
+    correct_netloc: String to overwrite the incoming URL's host and port with.
+      For example, this would be the "host.com:80" in "http://host.com:80/path."
+
+    Returns:
+      Patched FinalizeTransferUrl function.
+  """
+
+  def FinalizeTransferUrl(url):
+    modified_url = original_function(url)
+    parsed_url = urllib.parse.urlsplit(modified_url)
+    url_components = list(parsed_url)
```

Comment:
Since I have to do an "arbitrary index assignment" on the next line, I want to be as clear as possible before that.
I don't like arbitrary index assignments like "url_components[1] = ..." because "1" seems like a magic number to a reader.
Normally, the solution is to deconstruct, but there are 5+ items in `url_components`, and we only need to change one of them, so it's clunky to split it up like, `http, netloc, path, auth, params, ... = list(parsed_url)`.

I didn't combine lines because then I could use descriptive variables that makes it easier for the reader to follow what's going on. I hope that compensates a bit for the arbitrary index assignment if someone has to debug this later.